### ISSUE TITLE
Unsure how / no way to `replace` methods with `fakes`

### ISSUE THREAD

**Issue by akdor1154** [sinonjs/sinon#1850]
**Describe the bug**
If I fake a method (a function with a `this` reference) with `sinon.fake`, it looks like its `this` reference will become broken and unbindable.

This makes fakes entirely useless for stubbing methods of OOP style code.

**To Reproduce**
```js
const sinon = require('sinon')
const method = function() { return this.a; }
const fakeMethod = sinon.fake(method);

const o = {
  a: 'asdf',
  method: method,
  fakeMethod: fakeMethod,
  replacedMethod: method,
  explicitlyBoundMethod: method
}

sinon.replace(o, 'replacedMethod', sinon.fake(o.replacedMethod));
sinon.replace(o, 'explicitlyBoundMethod', sinon.fake(o.explicitlyBoundMethod).bind(o)); // i would hope this workaround would be considered unnecessary, even if it did work...

o.method() // 'asdf'
o.fakeMethod() // undefined
o.replacedMethod() // undefined
o.explicitlyBoundMethod() // undefined
```
**Expected behavior**
Setup as above.
```js
o.method() // 'asdf'
o.fakeMethod() // 'asdf'
o.replacedMethod() // 'asdf'
o.explicitlyBoundMethod() // 'asdf'
```

**Context (please complete the following information):**

* Library version: 6.0.1
* Environment: nodejs 8 on linux from nodesource

**Comment by mroderick** [comment]
I ran your example with `sinon@6.0.1` and current `master` in node 8 and node 10 on macOS.
All the lines returned `asdf` as desired.

Are you still able to reproduce this issue?

**Comment by akdor1154** [comment]
runkit repro: https://runkit.com/akdor1154/5b415c4a4999cd0012bb094d

**Comment by rgroothuijsen** [comment]
Confirmed this on my own machine (Ubuntu 18.04, Node 8.10.0). The explicitly bound method workaround does seem to produce the expected result if used in a slightly different way:

`sinon.fake(o.explicitlyBoundMethod.bind(o))`

**Comment by fatso83** [comment]
I verified this too, just replacing `console.log` with `assert` statements:

<details>
```
const assert = require("assert");
const sinon = require("sinon");
const method = function() {
  return this.a;
};
const fakeMethod = sinon.fake(method);

const o = {
  a: "asdf",
  method: method,
  fakeMethod: fakeMethod,
  replacedMethod: method,
  explicitlyBoundMethod: method
};

sinon.replace(o, "replacedMethod", sinon.fake(o.replacedMethod));
sinon.replace(
  o,
  "explicitlyBoundMethod",
  sinon.fake(o.explicitlyBoundMethod).bind(o)
); // i would hope this workaround would be considered unnecessary, even if it did work...

assert.equal(o.method(), "asdf");
assert.equal(o.fakeMethod(), undefined);
assert.equal(o.replacedMethod(), undefined);
assert.equal(o.explicitlyBoundMethod(), undefined);
```
</details>

**Comment by fatso83** [comment]
The real problem is that the fake doesn't return the value from the wrapped function. I trimmed down all the fat from the original:
<details>

```javascript
const assert = require("assert");
const sinon = require("sinon");
const method = sinon.stub().callsFake(function() {
  return this.foo;
});
const fakeMethod = sinon.fake(method);

const o = {
  foo: "asdf",
  method,
  fakeMethod
};

const results = o.fakeMethod();
assert.equal(method.callCount, 1);
assert.equal(fakeMethod.callCount, 1);
assert.equal(results, "asdf"); // fails with results being a sinon proxy
```
</details>

**Comment by fatso83** [comment]
OK, found a fix. It was the most obvious fix, but a bug in the driver code seemed to imply it was wrong. Submitting a PR.

**Comment by fatso83** [comment]
Published as 6.3.4
